dev/src/index.ts

```diff
@@ -703,6 +703,8 @@ export class Firestore {
    *   console.log(`Second document: ${JSON.stringify(docs[1])}`);
    * });
    */
+  getAll(...documentRefsOrReadOptions: Array<DocumentReference|ReadOptions>):
+      Promise<DocumentSnapshot[]>;
   getAll(
       documentRef: DocumentReference,
       ...moreDocumentRefsOrReadOptions: Array<DocumentReference|ReadOptions>):
```

Comment:
Interesting...  I thought the way overloads work in Typescript is that you list all the overloads and then you define the function so that its type is a superset of all of the overloads, but the function itself doesn't actually count as an overload.  So this would mean that:

1. This isn't actually defining multiple overloads, since the function implementation itself doesn't count as an overload.
2. I am surprised TypeScript even accepts this because the overload `getAll(...documentRefsOrReadOptions: Array<DocumentReference|ReadOptions>)` would accept `getAll()` or `getAll(readOptions)` but the actual function (`getAll(documentRef: DocumentReference, ...moreDocumentRefsOrReadOptions: Array<DocumentReference|ReadOptions>)` does not.  So the actual function definition is not a superset of all the overloads.

So I'm kinda' confused what is going on here.  But taking a step back I think overloads may not actually help us (sorry!).  I think the ideal function signature would be something like:

```
getAll(documentRef: DocumentReference, ...moreDocumentRefs: Array<DcoumentReference>, readOptions?: ReadOptions)
```

But TypeScript doesn't support this.  The next best signature is the one we have checked in right now, but that apparently breaks array destructuring (maybe because the array isn't guaranteed to have at least 1 element to match the leading `documentRef` argument?).

So I think that means that the best we can do is basically what you did originally and forego overloads and just make the signature:

```
getAll(...documentRefsOrReadOptions: Array<DocumentReference|ReadOptions>):
      Promise<DocumentSnapshot[]>;
```

This is overly permissive, since it allows `getAll()`, `getAll(readOptions)`, `getAll(doc, readOptions, doc, readOptions), etc. which are all invalid...  but I think it's the best we can do with TypeScript right now (there's an open issue that might help us if it's ever implemented: https://github.com/Microsoft/TypeScript/issues/1360)

Reply:
Thanks for the feedback. You're right, unfortunately that's the best we can do right now with what's available. I will update the PR with an alternative solution for the original signature that will be less permissive (won't allow `getAll()` at least). Let me know what you think.